staging/src/k8s.io/apiserver/pkg/cel/environment/base.go

```diff
@@ -95,6 +94,34 @@ var baseOpts = []VersionedOptions{
 			ext.Strings(ext.StringsVersion(2)),
 		},
 	},
+
+	// add set ext library in 1.29
+	{
+		IntroducedVersion: version.MajorMinor(1, 29),
+		EnvOptions: []cel.EnvOption{
+			ext.Sets(),
+		},
+	},
+	// add the new validator in 1.29
+	{
+		IntroducedVersion: version.MajorMinor(1, 29),
+		EnvOptions: []cel.EnvOption{
+			cel.ASTValidators(
+				cel.ValidateDurationLiterals(),
+				cel.ValidateTimestampLiterals(),
+				cel.ValidateRegexLiterals(),
+				cel.ValidateHomogeneousAggregateLiterals(),
```

Comment:
Let's put this above the string library section (`// String library`).  The new env options can be added together as a single 1.29 group under the above 1.28 group.

Reply:
I have moved this up and removed the removal of cel.HomogeneousAggregateLiterals() based on Tristian's comment. Please have a look when you have time. Thank you!